clio-client/src/main/scala/org/broadinstitute/clio/client/webclient/ClioWebClient.scala

```diff
@@ -98,57 +101,95 @@ class ClioWebClient(
     })(Keep.left)
     .run()
 
-  def dispatchRequest(request: HttpRequest): Future[HttpResponse] = {
+  def dispatchRequest(
+    request: HttpRequest,
+    credentials: Option[OAuth2Credentials]
+  ): Future[HttpResponse] = {
     val responsePromise = Promise[HttpResponse]()
-    requestStream.offer(request -> responsePromise).flatMap { queueResult =>
-      queueResult match {
-        case QueueOfferResult.Enqueued => ()
-        case QueueOfferResult.Dropped => {
-          responsePromise.failure {
-            new RuntimeException(
-              s"Client queue was too full to add the request: $request"
-            )
-          }
+    val requestWithCreds = credentials.fold(request) { creds =>
+      val now = Instant.now()
+
+      creds.synchronized {
+        /*
+         * Refresh the credential if either:
+         *   1. It has no underlying access token, or
+         *   2. The underlying access token has an expiration time, and it's expired
+         *
+         * DON'T refresh if the underlying access token has no expiration time,
+         * because refresh might not be supported for that token.
+         */
+        val maybeToken = Option(creds.getAccessToken)
+        lazy val maybeTokenExpirationInstant = for {
+          token <- maybeToken
+          expiration <- Option(token.getExpirationTime)
+        } yield {
+          expiration.toInstant
+        }
+        val shouldRefresh =
+          maybeToken.isEmpty || maybeTokenExpirationInstant.exists(
+            _.isAfter(now)
+          )
+
+        if (shouldRefresh) {
```

Comment:
Even if my other comment about refactoring the code is too hard to do now, I think it's worth moving the token refresh logic to another class instead of inlining it here.

This object would replace the use of `OAuth2Credentials` with an object that returns `HttpCredentials` on demand.